src/Microsoft.DocAsCode.Common/UriTemplates/UriTemplate.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DocAsCode.Common
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Text.RegularExpressions;
+
+    using Microsoft.DocAsCode.Plugins;
+
+    public class UriTemplate<T>
+    {
+        private static readonly Regex _marcoRegex = new Regex(@"{%\s*([\S]+?)\s*%}");
+        private static readonly Regex _pipelineRegex = new Regex(@"\|>\s*([^|\s]+)\s*(.*?)\s*(?:$|(?=\|>))");
+
+        private readonly string _template;
+        private readonly Func<string, T> _func;
+        private readonly IUriTemplatePipeline<T>[] _pipeline;
+        private readonly string[][] _parameters;
+
+        /// <summary>
+        /// step 1 (macro):
+        /// * env var:      {%abc%}                         -> expand env var abc.
+        /// Step 2 (pipeline):
+        /// * pipeline:     |> function parameters          -> "", and append a pipeline.
+        /// </summary>
+        public static UriTemplate<T> Parse(string template, Func<string, T> func, Func<string, IUriTemplatePipeline<T>> pipelineProvider)
+        {
+            if (template == null)
+            {
+                throw new ArgumentNullException(nameof(template));
+            }
+            if (func == null)
+            {
+                throw new ArgumentNullException(nameof(func));
+            }
+            if (pipelineProvider == null)
+            {
+                throw new ArgumentNullException(nameof(pipelineProvider));
+            }
+            var t = _marcoRegex.Replace(
+                template,
+                m => Environment.GetEnvironmentVariable(m.Groups[1].Value));
+            List<IUriTemplatePipeline<T>> pipeline = new List<IUriTemplatePipeline<T>>();
+            List<string[]> parameters = new List<string[]>();
+            t = _pipelineRegex.Replace(
+                t,
+                m =>
+                {
+                    pipeline.Add(pipelineProvider(m.Groups[1].Value));
```

Comment:
>pipeline.Add [](start = 20, length = 12)

seems the list will contain one item at most. No need to use a list?

Reply:
no, you can write multi-pipeline in one uri template, e.g.: `a|>x|>y`

Reply:
@superyyrrzz , see new test case.